File: 2026-09/src/day_11/leet_202/Solution2.java
package day_11.leet_202;

class Solution2 {
    public boolean isHappy(int n) {
        int slow = n;
        int fast = step(n);
        while (slow != fast) {
            slow = step(slow);
            fast = step(step(fast));
        }

        return slow == 1;
    }

    static int step(int x) {
        int sum = 0;
        while (x > 0) {
            int d = x % 10;
            sum += d * d;
            x /= 10;
        }

        return sum;
    }
}

File: 2026-09/src/day_11/leet_202/Solution2.py
class Solution:
    def isHappy(self, n: int) -> bool:

        def step(x):
            return sum(int(d) ** 2 for d in str(x))

        slow, fast = n, step(n)
        while slow != fast:
            slow = step(slow)
            fast = step(step(fast))

        return slow == 1
